File: peaky_finders/src/peaky_finders/serve/seek_plan.py
"""Persist goal-seek committed path under ``seek.plan`` in project YAML."""

from __future__ import annotations

from collections.abc import Mapping
from pathlib import Path
from typing import Any

from pydantic import ValidationError

from peaky_finders.core.preset import load_preset, normalize_site_tags, update_preset_yaml_tree
from peaky_finders.core.preset.model import SeekPlan, SeekPlanHop
from peaky_finders.serve.sites import site_row_from_yaml_ent, unique_site_slug, validate_site_name


class ServeSeekPlanError(Exception):
    """Goal-seek plan could not be loaded or saved."""


def _plan_hop_to_yaml(hop: SeekPlanHop) -> dict[str, Any]:
    if hop.site:
        return {"site": hop.site}
    row: dict[str, Any] = {"loc": [hop.loc[0], hop.loc[1]]}  # type: ignore[index]
    if hop.height_m is not None:
        row["height_m"] = float(hop.height_m)
    return row


def seek_plan_to_api(plan: SeekPlan | None) -> dict[str, Any] | None:
    if plan is None:
        return None
    return {
        "start": plan.start,
        "goal": [plan.goal[0], plan.goal[1]],
        "complete": bool(plan.complete),
        "hops": [_plan_hop_to_yaml(hop) for hop in plan.hops],
    }


def _normalize_plan_patch(body: Mapping[str, Any]) -> dict[str, Any]:
    if not isinstance(body, Mapping):
        raise ServeSeekPlanError("plan body must be an object")
    start = str(body.get("start", "")).strip()
    if not start:
        raise ServeSeekPlanError("start is required")
    goal_raw = body.get("goal")
    if not isinstance(goal_raw, (list, tuple)) or len(goal_raw) != 2:
        raise ServeSeekPlanError("goal must be [lat, lon]")
    hops_raw = body.get("hops")
    if not isinstance(hops_raw, list) or not hops_raw:
        raise ServeSeekPlanError("hops must be a non-empty list")
    hops: list[dict[str, Any]] = []
    for idx, item in enumerate(hops_raw):
        if not isinstance(item, Mapping):
            raise ServeSeekPlanError(f"hops[{idx}] must be an object")
        if item.get("site") is not None:
            site = str(item["site"]).strip()
            if not site:
                raise ServeSeekPlanError(f"hops[{idx}].site is required")
            hops.append({"site": site})
            continue
        loc = item.get("loc")
        if not isinstance(loc, (list, tuple)) or len(loc) != 2:
            raise ServeSeekPlanError(f"hops[{idx}] requires site or loc [lat, lon]")
        row: dict[str, Any] = {"loc": [float(loc[0]), float(loc[1])]}
        if item.get("height_m") is not None:
            row["height_m"] = float(item["height_m"])
        hops.append(row)
    return {
        "start": start,
        "goal": [float(goal_raw[0]), float(goal_raw[1])],
        "complete": bool(body.get("complete")),
        "hops": hops,
    }


def load_seek_plan_payload(preset_path: Path) -> dict[str, Any] | None:
    preset = load_preset(preset_path)
    return seek_plan_to_api(preset.seek.plan)


def patch_seek_plan(preset_path: Path, body: Mapping[str, Any]) -> dict[str, Any]:
    normalized = _normalize_plan_patch(body)
    plan = SeekPlan.model_validate(normalized)

    def mutator(_yaml_rt: Any, root: dict[str, Any]) -> dict[str, Any]:
        seek_raw = root.get("seek")
        if not isinstance(seek_raw, dict):
            seek_raw = {}
            root["seek"] = seek_raw
        seek_raw["plan"] = {
            "start": plan.start,
            "goal": [plan.goal[0], plan.goal[1]],
            "complete": plan.complete,
            "hops": [_plan_hop_to_yaml(hop) for hop in plan.hops],
        }
        return seek_raw["plan"]

    try:
        update_preset_yaml_tree(preset_path, mutator, validate=True)
    except ValidationError as e:
        raise ServeSeekPlanError(str(e)) from e
    except ValueError as e:
        raise ServeSeekPlanError(str(e)) from e
    api = seek_plan_to_api(plan)
    assert api is not None
    return api


def _loc_dedupe_key(loc: tuple[float, float], height_m: float | None) -> tuple[float, float, float | None]:
    hm = round(float(height_m), 1) if height_m is not None else None
    return (round(loc[0], 6), round(loc[1], 6), hm)


def _merge_tags_into_site_entry(
    sites_raw: dict[str, Any],
    slug: str,
    add_tags: list[str],
) -> dict[str, object] | None:
    """Merge add_tags onto an existing site; return API row when tags change."""
    ent = sites_raw.get(slug)
    if not isinstance(ent, dict):
        raise ValueError(f"site not found: {slug!r}")
    if "type" in ent:
        raise ValueError(f"sites.{slug}.type is removed; use tags")
    current = set(normalize_site_tags(ent.get("tags")))
    merged = current | set(add_tags)
    if merged == current:
        return None
    ent["tags"] = sorted(merged)
    return site_row_from_yaml_ent(slug, ent)


def _plan_path_site_slugs(start: str, hops: list[dict[str, Any]]) -> list[str]:
    slugs: list[str] = [start]
    seen = {start}
    for hop in hops:
        site = str(hop.get("site", "")).strip()
        if not site or site in seen:
            continue
        seen.add(site)
        slugs.append(site)
    return slugs


def convert_seek_plan_locs_to_sites(
    preset_path: Path,
    *,
    name_prefix: str,
    tags: list[str],
) -> dict[str, Any]:
    """Create preset sites from seek plan coordinate hops; rewrite hops as site refs."""
    prefix = validate_site_name(name_prefix)
    tags_value = normalize_site_tags(tags)
    if not tags_value:
        raise ServeSeekPlanError("tags must be a non-empty list")

    def mutator(_yaml_rt: Any, root: dict[str, Any]) -> dict[str, Any]:
        seek_raw = root.get("seek")
        if not isinstance(seek_raw, dict):
            raise ServeSeekPlanError("no seek plan")
        plan_raw = seek_raw.get("plan")
        if not isinstance(plan_raw, dict):
            raise ServeSeekPlanError("no seek plan")

        plan = SeekPlan.model_validate(plan_raw)
        if not any(hop.loc is not None for hop in plan.hops):
            raise ServeSeekPlanError("no coordinate hops to convert")

        sites_raw = root.get("sites")
        if sites_raw is None:
            sites_raw = {}
            root["sites"] = sites_raw
        if not isinstance(sites_raw, dict):
            raise ValueError("preset sites must be a mapping")

        existing = {str(k) for k in sites_raw.keys()}
        coord_to_slug: dict[tuple[float, float, float | None], str] = {}
        loc_counter = 0
        created_rows: list[dict[str, object]] = []
        new_hops: list[dict[str, Any]] = []

        for hop in plan.hops:
            if hop.site:
                new_hops.append({"site": hop.site})
                continue
            assert hop.loc is not None
            key = _loc_dedupe_key(hop.loc, hop.height_m)
            slug = coord_to_slug.get(key)
            if slug is None:
                loc_counter += 1
                name = f"{prefix} {loc_counter}"
                slug = unique_site_slug(existing, name)
                existing.add(slug)
                coord_to_slug[key] = slug
                entry: dict[str, Any] = {
                    "name": name,
                    "loc": [hop.loc[0], hop.loc[1]],
                    "tags": list(tags_value),
                }
                sites_raw[slug] = entry
                created_rows.append(site_row_from_yaml_ent(slug, entry))
            new_hops.append({"site": slug})

        seek_raw["plan"] = {
            "start": plan.start,
            "goal": [plan.goal[0], plan.goal[1]],
            "complete": plan.complete,
            "hops": new_hops,
        }
        api_plan = seek_plan_to_api(SeekPlan.model_validate(seek_raw["plan"]))
        assert api_plan is not None

        created_slugs = {str(row["slug"]) for row in created_rows}
        tagged_rows: list[dict[str, object]] = []
        for slug in _plan_path_site_slugs(plan.start, new_hops):
            if slug in created_slugs:
                continue
            row = _merge_tags_into_site_entry(sites_raw, slug, tags_value)
            if row is not None:
                tagged_rows.append(row)

        return {
            "sites": created_rows + tagged_rows,
            "plan": api_plan,
            "converted": len(created_rows),
            "tagged": len(tagged_rows),
        }

    try:
        result = update_preset_yaml_tree(preset_path, mutator, validate=True)
    except ValidationError as e:
        raise ServeSeekPlanError(str(e)) from e
    except ValueError as e:
        raise ServeSeekPlanError(str(e)) from e
    return dict(result)


def clear_seek_plan(preset_path: Path) -> None:
    def mutator(_yaml_rt: Any, root: dict[str, Any]) -> None:
        seek_raw = root.get("seek")
        if not isinstance(seek_raw, dict):
            return
        seek_raw.pop("plan", None)
        if not seek_raw:
            root.pop("seek", None)

    update_preset_yaml_tree(preset_path, mutator, validate=True, prune=False)

File: crates/peaky-serve/src/seek_plan.rs
//! Persist goal-seek committed path under ``seek.plan`` in project YAML.

use std::collections::{HashMap, HashSet};
use std::path::Path;

use anyhow::{Context, Result};
use peaky_preset::{
    insert_preset_site, load_preset, load_preset_raw, normalize_site_tags, patch_preset_sites_tags,
    site_row_from_entry, slugify, unique_site_slug, write_preset_document, SeekPlan, SeekPlanHop,
    SiteEntry,
};
use serde_json::{json, Value};
use serde_yaml::Mapping;

#[derive(Debug, thiserror::Error)]
#[error("{0}")]
pub struct SeekPlanError(pub String);

fn plan_hop_to_yaml(hop: &SeekPlanHop) -> Mapping {
    let mut row = Mapping::new();
    if let Some(site) = &hop.site {
        row.insert(
            serde_yaml::Value::from("site"),
            serde_yaml::Value::from(site.clone()),
        );
        return row;
    }
    if let Some(loc) = hop.loc {
        row.insert(
            serde_yaml::Value::from("loc"),
            serde_yaml::Value::from(vec![loc[0], loc[1]]),
        );
    }
    if let Some(h) = hop.height_m {
        row.insert(serde_yaml::Value::from("height_m"), serde_yaml::Value::from(h));
    }
    row
}

pub fn seek_plan_to_api(plan: &Option<SeekPlan>) -> Value {
    match plan {
        None => Value::Null,
        Some(plan) => json!({
            "start": plan.start,
            "goal": [plan.goal[0], plan.goal[1]],
            "complete": plan.complete,
            "hops": plan.hops.iter().map(plan_hop_to_json).collect::<Vec<_>>(),
        }),
    }
}

fn plan_hop_to_json(hop: &SeekPlanHop) -> Value {
    if let Some(site) = &hop.site {
        return json!({ "site": site });
    }
    let mut row = json!({ "loc": hop.loc });
    if let Some(h) = hop.height_m {
        row["height_m"] = json!(h);
    }
    row
}

pub fn load_seek_plan_payload(preset_path: &Path) -> Result<Value> {
    let preset = load_preset(preset_path)?;
    Ok(seek_plan_to_api(&preset.seek.plan))
}

pub fn patch_seek_plan(preset_path: &Path, body: &Value) -> Result<Value, SeekPlanError> {
    let plan: SeekPlan = serde_json::from_value(body.clone())
        .map_err(|e| SeekPlanError(format!("invalid seek plan: {e}")))?;
    if plan.start.trim().is_empty() {
        return Err(SeekPlanError("start is required".into()));
    }
    if plan.hops.is_empty() {
        return Err(SeekPlanError("hops must be a non-empty list".into()));
    }
    for hop in &plan.hops {
        if hop.site.is_none() && hop.loc.is_none() {
            return Err(SeekPlanError("each hop requires site or loc".into()));
        }
    }

    persist_seek_plan(preset_path, &plan)?;
    Ok(seek_plan_to_api(&Some(plan)))
}

fn persist_seek_plan(preset_path: &Path, plan: &SeekPlan) -> Result<(), SeekPlanError> {
    let mut raw = load_preset_raw(preset_path).map_err(|e| SeekPlanError(e.to_string()))?;
    let map = raw
        .as_mapping_mut()
        .ok_or_else(|| SeekPlanError("preset root must be a mapping".into()))?;
    let seek = map
        .entry(serde_yaml::Value::from("seek"))
        .or_insert_with(|| serde_yaml::Value::Mapping(Mapping::new()));
    if let serde_yaml::Value::Mapping(seek_map) = seek {
        seek_map.insert(
            serde_yaml::Value::from("plan"),
            serde_yaml::to_value(plan).map_err(|e| SeekPlanError(e.to_string()))?,
        );
    }
    write_preset_document(preset_path, map).map_err(|e| SeekPlanError(e.to_string()))?;
    Ok(())
}

pub fn clear_seek_plan(preset_path: &Path) -> Result<(), SeekPlanError> {
    let mut raw = load_preset_raw(preset_path).map_err(|e| SeekPlanError(e.to_string()))?;
    let map = raw.as_mapping_mut().ok_or_else(|| SeekPlanError("preset root must be a mapping".into()))?;
    if let Some(seek) = map.get_mut(&serde_yaml::Value::from("seek")) {
        if let serde_yaml::Value::Mapping(seek_map) = seek {
            seek_map.remove(&serde_yaml::Value::from("plan"));
            if seek_map.is_empty() {
                map.remove(&serde_yaml::Value::from("seek"));
            }
        }
    }
    write_preset_document(preset_path, map).map_err(|e| SeekPlanError(e.to_string()))?;
    Ok(())
}

fn loc_dedupe_key(loc: [f64; 2], height_m: Option<f64>) -> String {
    let hm = height_m.map(|h| (h * 10.0).round() / 10.0);
    format!(
        "{:.6},{:.6},{}",
        loc[0],
        loc[1],
        hm.map(|h| h.to_string()).unwrap_or_else(|| "null".into())
    )
}

/// Highest trailing number on an existing site whose name or slug matches ``prefix``.
fn max_existing_prefixed_site_number(
    sites: &HashMap<String, SiteEntry>,
    prefix: &str,
) -> u32 {
    let prefix = prefix.trim();
    if prefix.is_empty() {
        return 0;
    }
    let slug_base = slugify(prefix);
    let mut max_num = 0u32;
    for (slug, site) in sites {
        if let Some(n) = parse_prefixed_site_name_number(&site.name, prefix) {
            max_num = max_num.max(n);
        }
        if let Some(n) = parse_prefixed_site_slug_number(slug, &slug_base) {
            max_num = max_num.max(n);
        }
    }
    max_num
}

fn parse_prefixed_site_name_number(name: &str, prefix: &str) -> Option<u32> {
    let name = name.trim();
    let prefix = prefix.trim();
    if prefix.is_empty() || name.len() < prefix.len() {
        return None;
    }
    if !name[..prefix.len()].eq_ignore_ascii_case(prefix) {
        return None;
    }
    let rest = name[prefix.len()..].trim_start();
    if rest.is_empty() || !rest.chars().all(|c| c.is_ascii_digit()) {
        return None;
    }
    rest.parse().ok()
}

fn parse_prefixed_site_slug_number(slug: &str, slug_base: &str) -> Option<u32> {
    if slug_base.is_empty() || slug == slug_base {
        return None;
    }
    if !slug.starts_with(slug_base) {
        return None;
    }
    let mut rest = &slug[slug_base.len()..];
    if let Some(stripped) = rest.strip_prefix('-') {
        rest = stripped;
    }
    let digits: String = rest.chars().take_while(|c| c.is_ascii_digit()).collect();
    if digits.is_empty() {
        return None;
    }
    digits.parse().ok()
}

fn plan_path_site_slugs(start: &str, hops: &[Mapping]) -> Vec<String> {
    let mut slugs = vec![start.to_string()];
    let mut seen = HashSet::from([start.to_string()]);
    for hop in hops {
        if let Some(site) = hop.get(&serde_yaml::Value::from("site")).and_then(|v| v.as_str()) {
            let site = site.trim();
            if site.is_empty() || seen.contains(site) {
                continue;
            }
            seen.insert(site.to_string());
            slugs.push(site.to_string());
        }
    }
    slugs
}

pub fn convert_seek_plan_locs_to_sites(
    preset_path: &Path,
    name_prefix: &str,
    tags: &[String],
) -> Result<Value, SeekPlanError> {
    let prefix = name_prefix.trim();
    if prefix.is_empty() {
        return Err(SeekPlanError("name_prefix is required".into()));
    }
    let tags_value = normalize_site_tags(Some(&serde_yaml::Value::Sequence(
        tags.iter()
            .map(|t| serde_yaml::Value::from(t.clone()))
            .collect(),
    )))
    .map_err(|e| SeekPlanError(e))?;
    if tags_value.is_empty() {
        return Err(SeekPlanError("tags must be a non-empty list".into()));
    }

    let preset = load_preset(preset_path).map_err(|e| SeekPlanError(e.to_string()))?;
    let plan = preset
        .seek
        .plan
        .clone()
        .ok_or_else(|| SeekPlanError("no seek plan".into()))?;
    if !plan.hops.iter().any(|h| h.loc.is_some()) {
        return Err(SeekPlanError("no coordinate hops to convert".into()));
    }

    let mut preset = preset;
    let mut existing: HashSet<String> = preset.sites.keys().cloned().collect();
    let mut coord_to_slug: HashMap<String, String> = HashMap::new();
    let mut loc_counter = max_existing_prefixed_site_number(&preset.sites, prefix);
    let mut created_rows = Vec::new();
    let mut new_hops = Vec::new();

    for hop in &plan.hops {
        if let Some(site) = &hop.site {
            new_hops.push(SeekPlanHop {
                site: Some(site.clone()),
                loc: None,
                height_m: None,
            });
            continue;
        }
        let loc = hop.loc.context("hop loc").map_err(|e| SeekPlanError(e.to_string()))?;
        let key = loc_dedupe_key(loc, hop.height_m);
        let slug = if let Some(slug) = coord_to_slug.get(&key) {
            slug.clone()
        } else {
            loc_counter += 1;
            let name = format!("{prefix} {loc_counter}");
            let slug = unique_site_slug(&existing, &name);
            existing.insert(slug.clone());
            coord_to_slug.insert(key, slug.clone());
            let entry = SiteEntry {
                name,
                loc,
                tags: tags_value.clone(),
                height_m: hop.height_m,
                description: None,
            };
            preset.sites.insert(slug.clone(), entry.clone());
            created_rows.push(json!(site_row_from_entry(&slug, &entry)));
            slug
        };
        new_hops.push(SeekPlanHop {
            site: Some(slug),
            loc: None,
            height_m: None,
        });
    }

    let created_slug_list: Vec<String> = created_rows
        .iter()
        .filter_map(|row| row.get("slug").and_then(|v| v.as_str()).map(str::to_string))
        .collect();
    let created_slugs: HashSet<String> = created_slug_list.iter().cloned().collect();
    let mut tagged_rows = Vec::new();
    let yaml_hops: Vec<Mapping> = new_hops.iter().map(plan_hop_to_yaml).collect();
    for slug in plan_path_site_slugs(&plan.start, &yaml_hops) {
        if created_slugs.contains(&slug) {
            continue;
        }
        if let Some(site) = preset.sites.get_mut(&slug) {
            let current: HashSet<String> = site.tags.iter().cloned().collect();
            let merged: HashSet<String> = current.union(&tags_value.iter().cloned().collect()).cloned().collect();
            if merged != current {
                site.tags = merged.into_iter().collect::<Vec<_>>();
                site.tags.sort();
                tagged_rows.push(json!(site_row_from_entry(&slug, site)));
            }
        }
    }

    let path_slugs = plan_path_site_slugs(&plan.start, &yaml_hops);
    let tagged_slugs: Vec<String> = tagged_rows
        .iter()
        .filter_map(|row| row.get("slug").and_then(|v| v.as_str()).map(str::to_string))
        .collect();
    for slug in &created_slug_list {
        let entry = preset
            .sites
            .get(slug)
            .ok_or_else(|| SeekPlanError(format!("missing created site {slug}")))?;
        insert_preset_site(preset_path, slug, entry).map_err(|e| SeekPlanError(e.to_string()))?;
    }
    if !tagged_slugs.is_empty() {
        patch_preset_sites_tags(preset_path, &tagged_slugs, &tags_value, &[])
            .map_err(|e| SeekPlanError(e.to_string()))?;
    }

    let new_plan = SeekPlan {
        start: plan.start,
        goal: plan.goal,
        complete: plan.complete,
        hops: new_hops,
    };
    persist_seek_plan(preset_path, &new_plan)?;
    preset.seek.plan = Some(new_plan);

    let converted = created_rows.len();
    let tagged = tagged_rows.len();
    Ok(json!({
        "sites": created_rows.into_iter().chain(tagged_rows).collect::<Vec<_>>(),
        "created_slugs": created_slug_list,
        "path_slugs": path_slugs,
        "plan": seek_plan_to_api(&preset.seek.plan),
        "converted": converted,
        "tagged": tagged,
    }))
}

#[cfg(test)]
mod tests {
    use super::*;

    fn site(name: &str) -> SiteEntry {
        SiteEntry {
            name: name.into(),
            loc: [39.0, -119.0],
            tags: vec![],
            height_m: None,
            description: None,
        }
    }

    #[test]
    fn max_prefixed_number_from_names_and_slugs() {
        let mut sites = HashMap::new();
        sites.insert("relay1".into(), site("Relay 1"));
        sites.insert("relay5".into(), site("Relay 5"));
        sites.insert("relay1-2".into(), site("Relay 1"));
        sites.insert("fox-mountain".into(), site("Fox Mountain"));
        assert_eq!(max_existing_prefixed_site_number(&sites, "Relay"), 5);
        assert_eq!(max_existing_prefixed_site_number(&sites, "relay"), 5);
        assert_eq!(max_existing_prefixed_site_number(&sites, "Fox"), 0);
    }

    #[test]
    fn parse_prefixed_name_requires_numeric_suffix() {
        assert_eq!(parse_prefixed_site_name_number("Relay 12", "Relay"), Some(12));
        assert_eq!(parse_prefixed_site_name_number("relay 3", "Relay"), Some(3));
        assert!(parse_prefixed_site_name_number("Relay Peak", "Relay").is_none());
    }

    fn write_convert_fixture(dir: &std::path::Path) -> std::path::PathBuf {
        let config = r#"modem_presets:
  fixture-modem:
    frequency_mhz: 915.0
    bandwidth_khz: 125.0
    spreading_factor: 10
    coding_rate: 5
    implementation_margin_db: 3.0
    power_dbm: 22.0
    sensitivity_dbm: -132.0
environment_presets:
  fixture-desert:
    climate: desert
    polarization: vertical
    clutter_height_m: 1.0
    fresnel_clearance_fraction: 0.25
    coverage_pessimism_db: 0.0
    situation_pct: 95.0
    time_pct: 95.0
simulation:
  radius_km: 50.0
  modem: fixture-modem
  environment: fixture-desert
  transmitter: {height_m: 2.0, gain_dbi: 3.0, loss_db: 2.0}
  receiver: {height_m: 2.0, gain_dbi: 3.0, loss_db: 2.0}
seek:
  peak_bin_size_m: 1500
  max_candidates: 4
  plan:
    start: start
    goal: [40.25, -119.25]
    hops:
      - loc: [40.05, -119.45]
sites:
  start:
    name: Start
    loc: [40.0, -119.5]
    tags: [installed]
links: []
"#;
        let path = dir.join("config.yaml");
        std::fs::write(&path, config).expect("config");
        path
    }

    #[test]
    fn convert_loc_hop_becomes_site_and_keeps_path() {
        let tmp = tempfile::TempDir::new().expect("tempdir");
        let path = write_convert_fixture(tmp.path());
        let result =
            convert_seek_plan_locs_to_sites(&path, "Relay", &["planned".into()]).expect("convert");
        assert_eq!(result["converted"], 1);
        let created = result["created_slugs"]
            .as_array()
            .expect("created_slugs");
        assert_eq!(created.len(), 1);
        let slug = created[0].as_str().expect("slug");
        let path_slugs = result["path_slugs"].as_array().expect("path_slugs");
        assert!(path_slugs.iter().any(|v| v.as_str() == Some("start")));
        assert!(path_slugs.iter().any(|v| v.as_str() == Some(slug)));

        let preset = load_preset(&path).expect("reload");
        let site = preset.sites.get(slug).expect("created site");
        assert_eq!(site.loc, [40.05, -119.45]);
        assert!(site.tags.contains(&"planned".to_string()));
        let plan = preset.seek.plan.expect("plan");
        assert_eq!(plan.hops.len(), 1);
        assert_eq!(plan.hops[0].site.as_deref(), Some(slug));
        assert!(plan.hops[0].loc.is_none());
    }
}
